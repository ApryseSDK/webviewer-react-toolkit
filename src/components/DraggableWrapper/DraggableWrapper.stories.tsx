import React from 'react';
import { DndProvider } from 'react-dnd';
import Backend from 'react-dnd-html5-backend';
import DraggableWrapper from '../DraggableWrapper';
import docs from './README.md';

export default { title: 'DraggableWrapper', parameters: { info: docs } };

export const basic = () => (
  <DndProvider backend={Backend}>
    <DraggableWrapper index={0}>
      <div style={{ padding: 16, border: '1px solid red', background: 'lightgray' }}>This div is draggable</div>
    </DraggableWrapper>
  </DndProvider>
);
