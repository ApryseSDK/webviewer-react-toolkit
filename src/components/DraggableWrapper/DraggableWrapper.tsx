import classnames from 'classnames';
import React, { FC, ReactNode, useEffect, useRef, HTMLAttributes } from 'react';
import { DragObjectWithType, useDrag, useDrop } from 'react-dnd';

const ItemTypes = { Draggable: 'draggable' };

export interface DraggableWrapperProps extends HTMLAttributes<HTMLDivElement> {
  /**
   * The current index of the draggable wrapper.
   */
  index: number;
  /**
   * Call instead of providing children if you wish to use the isDragging prop.
   */
  onRenderChildren?: (isDragging: boolean) => ReactNode;
  /**
   * If given, will be called any time the draggable wrapper is moved.
   */
  onMove?: (fromIndex: number, toIndex: number) => void;
  /**
   * If given, will be called every time the draggable wrapper is dragging.
   * This **must** be memoized in order to work properly. If it is not, then
   * this will fire whenever the component updated regardless of whether drag
   * has started or not.
   */
  onDragStart?: () => void;
  /**
   * If given, will be called every time the draggable wrapper is not dragging.
   * This **must** be memoized in order to work properly. If it is not, then
   * this will fire whenever the component updated regardless of whether drag
   * has started or not.
   */
  onDragEnd?: () => void;
}

interface DragItem extends DragObjectWithType {
  index: number;
}

export const DraggableWrapper: FC<DraggableWrapperProps> = ({
  index,
  onMove,
  onRenderChildren,
  children,
  className,
  onDragStart,
  onDragEnd,
  ...divProps
}) => {
  const ref = useRef<HTMLDivElement>(null);

  const [, drop] = useDrop({
    accept: ItemTypes.Draggable,
    hover(item: DragItem) {
      // Cancel if ref hasn't been set up.
      if (!ref.current) return;

      // Previous index.
      const fromIndex = item.index;

      // Index it has been dragged to.
      const toIndex = index;

      // Cancel if index has not changed.
      if (fromIndex === toIndex) return;

      // Call onMove when index changes.
      onMove?.(fromIndex, toIndex);

      // Set the item index to be the new index.
      item.index = toIndex;
    },
  });

  const [{ isDragging }, drag] = useDrag({
    item: { type: ItemTypes.Draggable, index },
    collect: monitor => ({ isDragging: monitor.isDragging() }),
  });

  drag(drop(ref));

  useEffect(() => {
    if (isDragging) onDragStart?.();
    else onDragEnd?.();
  }, [isDragging, onDragStart, onDragEnd]);

  const draggableWrapperClass = classnames('ui__base ui__draggableWrapper', className);

  return (
    <>
      <div {...divProps} ref={ref} className={draggableWrapperClass}>
        {onRenderChildren ? onRenderChildren(isDragging) : children}
      </div>
    </>
  );
};
