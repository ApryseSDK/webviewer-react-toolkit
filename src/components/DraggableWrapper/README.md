This is a draggable wrapper that implements
[`react-dnd`](https://react-dnd.github.io/react-dnd) internally. It allows you
to have drag-and-drop-able elements. In this component library, it is used to
wrap the thumbnails of `PageOrganizer`.
