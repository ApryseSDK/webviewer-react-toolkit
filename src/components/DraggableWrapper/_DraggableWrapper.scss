.ui__draggableWrapper {
  display: inline-block;
}
