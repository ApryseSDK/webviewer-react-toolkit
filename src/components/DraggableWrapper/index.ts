export { DraggableWrapper as default, DraggableWrapperProps } from './DraggableWrapper';
