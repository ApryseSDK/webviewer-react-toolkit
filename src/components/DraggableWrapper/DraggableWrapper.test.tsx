import { mount } from 'enzyme';
import React from 'react';
import DraggableWrapper from '../DraggableWrapper';
import { DndProvider } from 'react-dnd';
import Backend from 'react-dnd-html5-backend';

describe('DraggableWrapper component', () => {
  it('renders its contents', () => {
    const draggableWrapper = mount(
      <DndProvider backend={Backend}>
        <DraggableWrapper index={0}>child</DraggableWrapper>
      </DndProvider>,
    );
    expect(draggableWrapper.find('.ui__draggableWrapper').length).toEqual(1);
  });

  it('snapshot renders default draggableWrapper', () => {
    const draggableWrapper = mount(
      <DndProvider backend={Backend}>
        <DraggableWrapper index={0}>child</DraggableWrapper>
      </DndProvider>,
    );
    expect(draggableWrapper).toMatchSnapshot();
  });
});
