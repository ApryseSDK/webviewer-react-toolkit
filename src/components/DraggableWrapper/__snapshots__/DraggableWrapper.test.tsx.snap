// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DraggableWrapper component snapshot renders default draggableWrapper 1`] = `
<DndProvider
  backend={[Function]}
>
  <Component
    index={0}
  >
    <div
      className="ui__base ui__draggableWrapper"
    >
      child
    </div>
  </Component>
</DndProvider>
`;
